File: v1/cmd/app/types/api/v1/attachment.go
package v1

import "time"

// AttachmentItem 单个附件详情。
type AttachmentItem struct {
	ID           int64     `json:"id"`
	Name         string    `json:"name"`
	OriginalName string    `json:"originalName"`
	URL          string    `json:"url"`
	ThumbURL     string    `json:"thumbUrl,omitempty"`
	Type         string    `json:"type"`
	MimeType     string    `json:"mimeType"`
	Size         int64     `json:"size"`
	Width        int       `json:"width,omitempty"`
	Height       int       `json:"height,omitempty"`
	ParentID     *int64    `json:"parentId,omitempty"`
	Variant      string    `json:"variant,omitempty"`
	GroupID      *int64    `json:"groupId,omitempty"`
	GroupName    string    `json:"groupName,omitempty"`
	RefCount     int64     `json:"refArticleCount"`
	CreatedAt    time.Time `json:"createdAt"`
}

// AttachmentListQuery 管理员附件列表查询。
type AttachmentListQuery struct {
	Page      int    `form:"page" binding:"omitempty,min=1"`
	PageSize  int    `form:"pageSize" binding:"omitempty,min=1,max=100"`
	Type      string `form:"type" binding:"omitempty,oneof=image document video audio other"`
	Keyword   string `form:"keyword" binding:"omitempty,max=200"`
	GroupID   *int64 `form:"groupId"`
	RootsOnly *bool  `form:"rootsOnly"` // 默认 true：仅根附件
}

// AttachmentListResponse 分页列表。
type AttachmentListResponse struct {
	Items    []AttachmentItem `json:"items"`
	Total    int64            `json:"total"`
	Page     int              `json:"page"`
	PageSize int              `json:"pageSize"`
}

// DeleteAttachmentResponse 删除结果。
type DeleteAttachmentResponse struct {
	ID int64 `json:"id"`
}

// UploadImageResponse 编辑器图片上传结果（仅含 url 和 alt）。
type UploadImageResponse struct {
	URL string `json:"url"`
	Alt string `json:"alt"`
}

// AttachmentFamilyArticleRef 引用家族内任一附件的文章。
type AttachmentFamilyArticleRef struct {
	ArticleID int64  `json:"articleId"`
	Title     string `json:"title"`
	Slug      string `json:"slug"`
}

// AttachmentMemberArticleRefs 单个家族成员在 article_attachments 中关联的文章列表。
type AttachmentMemberArticleRefs struct {
	AttachmentID int64                        `json:"attachmentId"`
	Articles     []AttachmentFamilyArticleRef `json:"articles"`
}

// AttachmentFamilyResponse 根附件及其派生子附件。
type AttachmentFamilyResponse struct {
	Root             AttachmentItem                `json:"root"`
	Children         []AttachmentItem              `json:"children"`
	MemberReferences []AttachmentMemberArticleRefs `json:"memberReferences"`
}

// AttachmentProcessingSettings 全局默认处理参数（存 settings 表 JSON）。
type AttachmentProcessingSettings struct {
	DefaultQuality int      `json:"defaultQuality"`
	AllowedFormats []string `json:"allowedFormats"`
}

// ProcessAttachmentRequest 从根附件生成派生子附件（压缩质量 + 可选输出格式）。
type ProcessAttachmentRequest struct {
	Quality int    `json:"quality" binding:"omitempty,min=1,max=100"`
	Format  string `json:"format" binding:"omitempty,oneof=jpeg jpg png gif webp ico"`
}

// UpdateAttachmentRequest 更新附件：可选物理重命名（移动磁盘文件、更新 path/url/name/original_name，并替换正文中的旧 URL）与可选分类。
// 至少提供 name 或 groupId 之一。groupId 省略表示不改分类；0 或负数表示清除分类；正数为有效分类 ID。根附件变更分类时，子附件 group_id 与根一致。
type UpdateAttachmentRequest struct {
	Name    *string `json:"name,omitempty" binding:"omitempty,min=1,max=255"`
	GroupID *int64  `json:"groupId,omitempty"`
}

// ReplaceAttachmentInArticlesRequest 将正文中 from 附件 URL 批量替换为 to。
type ReplaceAttachmentInArticlesRequest struct {
	FromAttachmentID int64   `json:"fromAttachmentId" binding:"required"`
	ToAttachmentID   int64   `json:"toAttachmentId" binding:"required"`
	ArticleIDs       []int64 `json:"articleIds" binding:"required,min=1"`
}

// ReplaceAttachmentInArticlesResponse 替换结果。
type ReplaceAttachmentInArticlesResponse struct {
	Updated int `json:"updated"`
}

// AttachmentGroupItem 附件分类（独立于文章分类）。
type AttachmentGroupItem struct {
	ID        int64  `json:"id"`
	Name      string `json:"name"`
	ParentID  *int64 `json:"parentId,omitempty"`
	SortOrder int    `json:"sortOrder"`
}

// CreateAttachmentGroupRequest 创建附件分类。
type CreateAttachmentGroupRequest struct {
	Name      string `json:"name" binding:"required,min=1,max=50"`
	ParentID  *int64 `json:"parentId"`
	SortOrder int    `json:"sortOrder" binding:"omitempty,min=0,max=1000000"`
}

// UpdateAttachmentGroupRequest 更新附件分类。
type UpdateAttachmentGroupRequest struct {
	Name      *string `json:"name" binding:"omitempty,min=1,max=50"`
	ParentID  *int64  `json:"parentId"`
	SortOrder *int    `json:"sortOrder" binding:"omitempty,min=0,max=1000000"`
}

File: v1/ui/admin/src/api/attachment.ts
import request from '../utils/request';

export interface Attachment {
  id: number;
  name: string;
  originalName: string;
  url: string;
  thumbUrl?: string;
  type: string;
  mimeType: string;
  size: number;
  width?: number;
  height?: number;
  parentId?: number | null;
  variant?: string;
  groupId?: number | null;
  groupName?: string;
  refArticleCount?: number;
  createdAt: string;
}

export interface AttachmentListParams {
  page?: number;
  pageSize?: number;
  keyword?: string;
  type?: string;
  groupId?: number;
  rootsOnly?: boolean;
}

export interface AttachmentListResponse {
  items: Attachment[];
  total: number;
  page: number;
  pageSize: number;
}

export interface AttachmentFamilyArticleRef {
  articleId: number;
  title: string;
  slug: string;
}

export interface AttachmentMemberArticleRefs {
  attachmentId: number;
  articles: AttachmentFamilyArticleRef[];
}

export interface AttachmentFamilyResponse {
  root: Attachment;
  children: Attachment[];
  memberReferences: AttachmentMemberArticleRefs[];
}

export interface AttachmentProcessingSettings {
  defaultQuality: number;
  allowedFormats: string[];
}

export type ProcessOutputFormat = 'jpeg' | 'png' | 'gif' | 'webp' | 'ico';

export interface ProcessAttachmentRequest {
  quality?: number;
  /** 省略或与源相同：不传 format 字段 */
  format?: ProcessOutputFormat | '';
}

export interface ReplaceAttachmentInArticlesRequest {
  fromAttachmentId: number;
  toAttachmentId: number;
  articleIds: number[];
}

export interface ReplaceAttachmentInArticlesResponse {
  updated: number;
}

export interface AttachmentGroupItem {
  id: number;
  name: string;
  parentId?: number | null;
  sortOrder: number;
}

export const getAttachments = (params?: AttachmentListParams) => {
  return request.get<any, { code: number; message: string; data: AttachmentListResponse }>(
    '/api/v1/admin/attachments',
    { params },
  );
};

export const deleteAttachment = (id: number) => {
  return request.delete<any, any>(`/api/v1/admin/attachments/${id}`);
};

export const uploadAttachment = (file: File, groupId?: number) => {
  const formData = new FormData();
  formData.append('file', file);
  const params = groupId != null && groupId > 0 ? { groupId } : {};
  return request.post<any, { code: number; message: string; data: Attachment }>(
    '/api/v1/admin/attachments/upload',
    formData,
    {
      params,
      headers: { 'Content-Type': 'multipart/form-data' },
    },
  );
};

export const getAttachmentFamily = (id: number) => {
  return request.get<any, { code: number; message: string; data: AttachmentFamilyResponse }>(
    `/api/v1/admin/attachments/${id}/family`,
  );
};

export const processAttachment = (id: number, body: ProcessAttachmentRequest) => {
  const payload: Record<string, unknown> = {};
  if (body.quality != null) payload.quality = body.quality;
  if (body.format) payload.format = body.format;
  return request.post<any, { code: number; message: string; data: Attachment }>(
    `/api/v1/admin/attachments/${id}/process`,
    payload,
  );
};

/** 至少提供 name 或 groupId 之一。groupId 为 0 表示清除分类。 */
export interface UpdateAttachmentRequest {
  name?: string;
  groupId?: number;
}

export const updateAttachment = (id: number, body: UpdateAttachmentRequest) => {
  return request.put<any, { code: number; message: string; data: Attachment }>(
    `/api/v1/admin/attachments/${id}`,
    body,
  );
};

export const replaceAttachmentInArticles = (body: ReplaceAttachmentInArticlesRequest) => {
  return request.post<any, { code: number; message: string; data: ReplaceAttachmentInArticlesResponse }>(
    '/api/v1/admin/attachments/replace-in-articles',
    body,
  );
};

export const getAttachmentProcessingSettings = () => {
  return request.get<any, { code: number; message: string; data: AttachmentProcessingSettings }>(
    '/api/v1/admin/attachments/settings',
  );
};

export const putAttachmentProcessingSettings = (body: AttachmentProcessingSettings) => {
  return request.put<any, { code: number; message: string; data: AttachmentProcessingSettings }>(
    '/api/v1/admin/attachments/settings',
    body,
  );
};

export const listAttachmentGroups = () => {
  return request.get<any, { code: number; message: string; data: AttachmentGroupItem[] }>(
    '/api/v1/admin/attachment-groups',
  );
};

export const createAttachmentGroup = (body: { name: string; parentId?: number | null; sortOrder?: number }) => {
  return request.post<any, { code: number; message: string; data: AttachmentGroupItem }>(
    '/api/v1/admin/attachment-groups',
    body,
  );
};

export const deleteAttachmentGroup = (id: number) => {
  return request.delete<any, { code: number; message: string; data: { id: number } }>(
    `/api/v1/admin/attachment-groups/${id}`,
  );
};
